ExPlat: Clean up LocalStorage on startup of ExPlat client

is4ac opened the pull request:
#### Changes proposed in this Pull Request

Currently, old ExperimentAssignments that have been stored in LocalStorage will continue to exist even after the experiment is over. The ExPlat team is concerned that the extra items in LocalStorage may be contributing to client errors related to exceeding the LocalStorage limit (though these are relatively rare).

* Removes all expired ExperimentAssignments that are stored in LocalStorage when the ExPlat client is initialized
* Added functions to the LocalStorage polyfill for testing purposes
* Added unit test for the new `removeExpiredExperimentAssignments` function

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Added unit test for `experiment-assignment-store.ts` in the `explat-client` package

Addresses [this ExPlat issue](https://github.com/Automattic/experimentation-platform/issues/622)

is4ac committed:
Clean up LocalStorage on startup of ExPlat client

- Removes all expired ExperimentAssignments that are stored in LocalStorage
- Added functions to the LocalStorage polyfill for testing purposes
- Added unit test for the new removeExpiredExperimentAssignments function

jessie-ross commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -93,6 +94,9 @@ export function createExPlatClient( config: Config ): ExPlatClient {
 	};
 
+	// Clean up LocalStorage on start up
+	removeExpiredExperimentAssignments();
```
Let's wrap this in a try-catch with error logging - we don't want any errors in this code base bubbling out.

is4ac replied:
That makes sense!

jessie-ross commented on packages/explat-client/src/internal/experiment-assignment-store.ts:
```diff
@@ -50,3 +51,26 @@ export function retrieveExperimentAssignment(
+ * Removes all expired and invalid experiment assignments in LocalStorage.
+ */
+export function removeExpiredExperimentAssignments(): void {
+	for ( let i = localStorage.length - 1; i >= 0; i-- ) {
```
Let's use list abstractions to make this code simpler - `map` and `filter`

is4ac replied:
Okay, yeah, that works. I wasn't sure what the general consensus was on using `map`/`filter` vs `for` loops

is4ac replied:
Actually, digging into this a bit more, looks like the only way to generate a list of `localStorage` keys to begin with would be to use a `for` loop (there is no built-in `localStorage` method to return a list of all keys, unless we used `Object.keys(localStorage)`, in which case, I'm really unsure how to create a polyfill for that). 

With that in mind, should I go ahead and refactor by using a for loop to generate a list of keys, and then do the list abstractions? Or should I leave it as one for loop for efficiency reasons?

jessie-ross replied:
> With that in mind, should I go ahead and refactor by using a for loop to generate a list of keys, and then do the list abstractions? 

The key here (and to replace any for loop that generates a list of numbers) is using a "range" function that generates a range of indexes, [since we don't have lodash we use this method](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_range): `[...Array(i).keys()] === [0, ..., i - 1]`

I would actually put that into a function: 
```js
const range = (i) => [...Array(i).keys()]
```

I accidentally wrote the whole thing out:
```js
range(localStorage.length)
    .map(i => localStorage.key(i))
    .filter(key => key?.startsWith(localStorageExperimentAssignmentKeyPrefix))
    .map(key => key.slice( localStorageExperimentAssignmentKeyPrefix.length + 1 ))
    .filter(experimentName => {
        try { 
            if ( ExperimentAssignments.isAlive(retrieveExperimentAssignment(experimentName)) }) ) 
                return false;
        } catch (_) { }
        return true;
    })
    .map(experimentName => localStorage.removeItem(localStorageExperimentAssignmentKey(experimentName))
```

>  I wasn't sure what the general consensus was on using map/filter vs for loops

The advantages of using list abstractions here (and almost always):
- It prevents most off-by-one errors.
- In between each step we have some very nice intermediate state we can use to visually ensure everything is working as expected.
- Adds simplicity by ensuring immutable data transforms and no access to previous state - all each step needs to know about is the single value it's given - not the code before or after it.
- Because of the above each of these steps are composable and reusable as-is - it is immediate that between break any two lines apart as functions, as well as the individual transforms.
- Because of the above we can see some useful ways of splitting this up into reusable, and _easily unit-testable_, functions: `getAllLocalStorageKeys`, `isLocalStorageExperimentAssignmentKey`. `experimentNameFromLocalStorageExperimentAssignmentKey`, with our main function becoming:

  ```js
  getAllLocalStorageKeys()
      .filter(isLocalStorageExperimentAssignmentKey)
      .map(experimentNameFromLocalStorageExperimentAssignmentKey)
      .filter(experimentName => {
          try { 
              if ( ExperimentAssignments.isAlive(retrieveExperimentAssignment(experimentName)) }) ) 
                  return false;
          } catch (_) { }
          return true;
      })
      .map(localStorageExperimentAssignmentKey)
      .map(key => localStorage.removeItem(key))
  ```

   Which may or may not be an improvement over the original code.

> Or should I leave it as one for loop for efficiency reasons?

It is always important to consider if we need performance optimisations before thinking about changing the code for performance reasons. 

So long as we keep an eye on algorithmic complexity (`O(n)` here), the size of `n` (only need to start considering it in the `10,000`-`100,000` range), and if there are any expensive operations going on (perhaps `localStorage` interactions), it is almost always the case we should make choices favouring all other factors over performance.

It is, however, important to know and recognise those situations where optimisation is necessary.

jessie-ross replied:
On another topic, it seems like we also have behaviour here where we remove invalidly (as in "validation") stored experiment assignments, I think we should be testing that too.

is4ac replied:
Thanks for the detailed and highly educational feedback! I will go ahead and make those changes.

> On another topic, it seems like we also have behaviour here where we remove invalidly (as in "validation") stored experiment assignments, I think we should be testing that too.

Yes, that's right. I'll add unit tests for that as well.

jessie-ross commented on packages/explat-client/src/internal/local-storage.ts:
```diff
@@ -20,6 +20,13 @@ const localStorage =
+				},
+				key: function ( index: number ): string | undefined {
+					const arr = this._data.keys;
+					return arr[ index ];
```
I'm not sure what's going on here?

Where does `keys` exist?

What's with the use of `arr`?

jessie-ross replied:
Did you test the polyfill?

is4ac replied:
> I'm not sure what's going on here?
> 
> Where does `keys` exist?
> 
> What's with the use of `arr`?

Oh, good catch. I meant to use `Object.keys(this._data)`.

And are you asking why the intermediate variable? `arr` is a bad name, I could change it to `keys`, or avoid the intermediate variable altogether and directly return `Object.keys(this._data)[ index ]`. I just thought it would be easier to read with the intermediate variable.

is4ac replied:
> Did you test the polyfill?

I guess not... oops. I just looked around the codebase, and I couldn't find any existing tests for the current polyfill. Should I create a new unit test file that tests all of the polyfill functions?

jessie-ross replied:
> And are you asking why the intermediate variable? arr is a bad name, I could change it to keys, or avoid the intermediate variable altogether and directly return Object.keys(this._data)[ index ]. I just thought it would be easier to read with the intermediate variable.

To me it decreases readability as now I have two lines and an extra variable, I have to look up the variable to see what it does. `Object.keys(this._data)[ index ]` should be very understandable to experienced developers.

> I guess not... oops. I just looked around the codebase, and I couldn't find any existing tests for the current polyfill. Should I create a new unit test file that tests all of the polyfill functions?

Yeah the story with the polyfill is that the code comes from a well battle-tested source, and the same time it should rarely if ever be needed. But if we are extending it, I think we should test the extensions.

is4ac replied:
> To me it decreases readability as now I have two lines and an extra variable, I have to look up the variable to see what it does. Object.keys(this._data)[ index ] should be very understandable to experienced developers.

Sure, that makes sense.

> Yeah the story with the polyfill is that the code comes from a well battle-tested source, and the same time it should rarely if ever be needed. But if we are extending it, I think we should test the extensions.

Gotcha, so I'll make a new test file that tests the extensions in that case.

jessie-ross commented on packages/explat-client/src/internal/test/experiment-assignment-store.ts:
```diff
@@ -55,4 +56,32 @@ describe( 'experiment-assignment-store', () => {
+		storeExperimentAssignment( validExperimentAssignment );
+		removeExpiredExperimentAssignments();
+		expect( localStorage.length ).toBe( 1 );
+	} );
```
Let's also have a test with combined alive/dead in storage.

is4ac replied:
What do you mean by combined alive/dead? Isn't this already what I do with this second `expect`?

jessie-ross replied:
Ah you're right, I missed the second store.

is4ac committed:
Address Jessie's comments.

- Added try-catch to removeExpiredExperimentAssignments call in create-explat-client
- Uses map and filter in removeExpiredExperimentAssignments now
- Added helper functions for readability
- Fixed key function in localStorage polyfill

is4ac committed:
Add tests for local-storage polyfill and experiment-assignment-store

jessie-ross commented on packages/explat-client/src/internal/experiment-assignment-store.ts:
```diff
@@ -53,24 +53,52 @@ export function retrieveExperimentAssignment(
-	for ( let i = localStorage.length - 1; i >= 0; i-- ) {
-		const key = localStorage.key( i );
+export function getAllLocalStorageKeys(): string[] {
+	const range = ( i: number ) => [ ...Array( i ).keys() ];
```
Let's move range out of here

jessie-ross commented on packages/explat-client/src/internal/experiment-assignment-store.ts:
```diff
@@ -1,10 +1,11 @@
+export const localStorageExperimentAssignmentKeyPrefix = 'explat-experiment-';
 
-const localStorageExperimentAssignmentKey = ( experimentName: string ): string =>
+export const localStorageExperimentAssignmentKey = ( experimentName: string ): string =>
```
Let's mention that these are exported only for testing (including all the helper functions below.)

If they are being exported only for testing and are simple functions they also don't need documentation generally.

is4ac replied:
Sounds good! Ironically, I actually had it commented like that at first and then second guessed myself and deleted all those comments at the last minute before committing 😅😂

jessie-ross reviewed (approved):
See minor comments, but LGTM 🚢 
Thanks for your work here!

(I will need to merge and deploy it.)

is4ac committed:
Fix documentation and address Jessie's comments

is4ac commented:
Great! I think I got everything fixed and committed now. Let me know if there are any other issues!

jessie-ross committed:
Change comments to JSDoc

jessie-ross commented:
I've just changed the "Export only" comments to use JSDoc so they are visible in highlights.
